### ISSUE TITLE
Add an Endpoint API

### ISSUE THREAD

**Issue by ericnicolaas** [Charitable/Charitable#306]
Related to #144 and possibly needing to be in place before that issue is completed. 

Currently, if you want to register a new endpoint in Charitable, you have to add a stack of functions, hook them up to the right hooks and optionally add new settings in the admin to allow people to modify the page to use for a particular endpoint. There are two many disparate components, and it's a hassle to register a new endpoint. It's also inefficient to do basic boolean checks on whether you're viewing a particular page.

For example, to look at what we're doing to set up the `forgot_password` endpoint: 

- `add_rewrite_endpoint()` in `Charitable_Post_Types::add_endpoints()` 
- `add_rewrite_rule()` in `Charitable_Post_Types::add_rewrite_rule()` 
- `charitable_get_forgot_password_page_permalink()` to get the permalink
- `charitable_is_forgot_password_page()` to check if we're on the forgot password page
- `Charitable_Templates::get_forgot_password_template()` to set up the template to display the content
- `charitable_template_forgot_password_content()` to load the page's template file
- `templates/account/forgot-password.php` to actually display the content

Currently, this is all split over six different files and it's very easy to get lost. If this were a page that can be modified via the admin settings, then you'd need to touch another file as well. 

Ideally, I'd like to replace this with a system that requires touching in just two or three places: 

1. A class that does the following: add the rewrite endpoint and rewrite rule; has a `get_page_permalink()` getter to retrieve the permalink for the endpoint; has an `is_page()` method to do the boolean check; has a `get_page_template()` method to get the template; and maybe has a `render_content()` method to load the content template.
2. A content template inside the `templates` directory.
3. A function call somewhere that registers the endpoint inside Charitable.

Registering your endpoint should be a matter of passing the endpoint ID and the corresponding endpoint object. Or maybe just the endpoint object: 

```
$endpoint = new Charitable_Endpoint_Forgot_Password();
charitable_register_endpoint( $endpoint );
```

This would need to happen early on, at least before `init`. Maybe in `after_setup_theme`?

`charitable_register_endpoint()` can be a wrapper around a class we use as a registry for all the endpoints. By using a registry, we should also be able to avoid some of the inefficiency we currently have when doing boolean checks (these are done repeatedly for the same endpoint, each time executing the same function).

`charitable_get_permalink()` and `charitable_is_page()` can also wrap around the endpoint registry class.

The individual endpoint functions should then be deprecated and call the relevant methods in the registry class.

**Comment by ericnicolaas** [comment]
Merged. Closed by 8222d8b

**Comment by ericnicolaas** [comment]
Just jotting down some usage notes here for developers to be documented.

There are two ways to get the Endpoint API object:

```
$api = charitable_get_helper( 'endpoints' );

$api = charitable()->endpoints();
```

Similarly, you can register an endpoint in a couple ways:

```
$endpoint = new Charitable_Endpoint_Forgot_Password();
charitable_register_endpoint( $endpoint ); 

$endpoint = new Charitable_Endpoint_Forgot_Password();
charitable()->endpoints()->register( $endpoint );
```

The URL for a particular endpoint can be retrieved in two ways:

```
$url = charitable_get_permalink( 'profile' );

$url = charitable()->endpoints()->get_page_url( 'profile' );
```

You can also check if you are on a particular endpoint in one of two ways:

```
$is_page = charitable_is_page( 'profile' );

$is_page = charitable()->endpoints()->is_page( 'profile' );
```

When using `get_page_url`, some endpoints require additional arguments, which can be passed in a second parameter as an array.

```
// Get a particular campaign's permalink
$url = charitable_get_permalink( 'campaign', array( 'campaign_id' => 123 ) );

// Get a particular campaign's donation page
$url = charitable_get_permalink( 'campaign_donation', array( 'campaign_id' => 123 ) );

// Get the cancellation url for a donation
$url = charitable_get_permalink( 'donation_cancellation', array( 'donation_id' => 123 ) );

// Get the processing page for a donation
$url = charitable_get_permalink( 'donation_processing', array( 'donation_id' => 123 ) );

// Get the receipt for a donation
$url = charitable_get_permalink( 'donation_receipt', array( 'donation_id' => 123 ) );

// Get the preview for an email
$url = charitable_get_permalink( 'email_preview', array( 'email_id' => 'new_donation' ) );
```

Finally, to get the current endpoint:

```
$endpoint = charitable_get_helper( 'endpoints' )->get_current_endpoint();

$endpoint = charitable()->endpoints()->get_current_endpoint();
```

As you can see, there are basically always two ways of interacting with the Endpoints API. Internally, we use the `charitable()->endpoints()` approach to fetching the object when we are registering a new endpoint, but we use `charitable_get_permalink()`and `charitable_is_page()` rather than their more verbose alternatives. 

Note also that `charitable_get_permalink()` and `charitable_is_page()` preceded the Endpoints API and have both existed since version 1.0.0.
